terraform-provider-tanzu-mission-control_v1.1.2.exe plugin crashed

### Describe the bug

│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-tanzu-mission-control_v1.1.2.exe plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x222b450]

goroutine 51 [running]:
github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/cluster.dataSourceTMCClusterRead({0x27eddb8, 0xc0001925a0}, 0xc00043
5c80, {0x24562e0?, 0xc0001d31c0?})
        github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/cluster/data_source_cluster.go:136 +0xdb0
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc000203ce0, {0x27eddf0, 0xc0005fe450}, 0xd?, {0x24562e0, 0xc0001d31c
0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/helper/schema/resource.go:359 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc000203ce0, {0x27eddf0, 0xc0005fe450}, 0xc000435c00, {0x245
62e0, 0xc0001d31c0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/helper/schema/resource.go:578 +0x145
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc0002cf710, {0x27edd48?, 0xc000363740?}, 0xc0002
ad780)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.12.0/helper/schema/grpc_provider.go:1179 +0x357
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc000269360, {0x27eddf0?, 0xc000483980?}, 0xc000080f00)
        github.com/hashicorp/terraform-plugin-go@v0.8.0/tfprotov5/tf5server/server.go:657 +0x409
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler({0x24b2520?, 0xc000269360}, {0x27eddf0, 0xc
000483980}, 0xc00007be00, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.8.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:421 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001b5a40, {0x27f14c0, 0xc00008a1a0}, 0xc0005f4ea0, 0xc0003c9c80, 0x333c0b0, 0x0)
        google.golang.org/grpc@v1.45.0/server.go:1282 +0xccf
google.golang.org/grpc.(*Server).handleStream(0xc0001b5a40, {0x27f14c0, 0xc00008a1a0}, 0xc0005f4ea0, 0x0)
        google.golang.org/grpc@v1.45.0/server.go:1619 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.45.0/server.go:921 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.45.0/server.go:919 +0x28a

Error: The terraform-provider-tanzu-mission-control_v1.1.2.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.


### Reproduction steps

1. terraform plan


### Expected behavior

..

### Additional context

_No response_

--- a/internal/resources/cluster/data_source_cluster.go
+++ b/internal/resources/cluster/data_source_cluster.go
@@ -127,20 +127,29 @@ func dataSourceTMCClusterRead(ctx context.Context, d *schema.ResourceData, m int
 		return diag.FromErr(err)
 	}
 
-	clusterSpec := constructSpec(d)
-
-	switch {
-	case resp.Cluster.Spec.TkgAws != nil:
-		resp.Cluster.Spec.TkgAws.Topology.NodePools = clusterSpec.TkgAws.Topology.NodePools
-	case resp.Cluster.Spec.TkgServiceVsphere != nil:
-		resp.Cluster.Spec.TkgServiceVsphere.Topology.NodePools = clusterSpec.TkgServiceVsphere.Topology.NodePools
-	case resp.Cluster.Spec.TkgVsphere != nil:
-		resp.Cluster.Spec.TkgVsphere.Topology.NodePools = clusterSpec.TkgVsphere.Topology.NodePools
-	}
+	setNodepoolForClusterResource(resp.Cluster.Spec, constructSpec(d))
 
 	if err := d.Set(SpecKey, flattenSpec(resp.Cluster.Spec)); err != nil {
 		return diag.FromErr(err)
 	}
 
 	return diags
+}
+
+func setNodepoolForClusterResource(respSpec, clusterSpec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec) {
+	if clusterSpec == nil {
+		return
+	}
+
+	switch {
+	// check for data-source read because only full name will be provided in terraform state file [spec will be empty].
+	case clusterSpec.TkgAws == nil && clusterSpec.TkgServiceVsphere == nil && clusterSpec.TkgVsphere == nil:
+		return
+	case clusterSpec.TkgAws != nil && clusterSpec.TkgAws.Topology != nil && respSpec.TkgAws.Topology != nil:
+		respSpec.TkgAws.Topology.NodePools = clusterSpec.TkgAws.Topology.NodePools
+	case clusterSpec.TkgServiceVsphere != nil && clusterSpec.TkgServiceVsphere.Topology != nil && respSpec.TkgServiceVsphere.Topology != nil:
+		respSpec.TkgServiceVsphere.Topology.NodePools = clusterSpec.TkgServiceVsphere.Topology.NodePools
+	case clusterSpec.TkgVsphere != nil && clusterSpec.TkgVsphere.Topology != nil && respSpec.TkgVsphere.Topology != nil:
+		respSpec.TkgVsphere.Topology.NodePools = clusterSpec.TkgVsphere.Topology.NodePools
+	}
 }

--- a/internal/resources/cluster/get_nodepool_test.go
+++ b/internal/resources/cluster/get_nodepool_test.go
@@ -1,0 +1,454 @@
+/*
+Copyright © 2022 VMware, Inc. All Rights Reserved.
+SPDX-License-Identifier: MPL-2.0
+*/
+
+package cluster
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	clustermodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster"
+	nodepoolmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster/nodepool"
+	tkgawsmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster/tkgaws"
+	tkgservicevspheremodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster/tkgservicevsphere"
+	tkgvspheremodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster/tkgvsphere"
+)
+
+func TestGetNodepoolForCluster(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		description      string
+		inputRespSpec    *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec
+		inputClusterSpec *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec
+		expectedSpec     *clustermodel.VmwareTanzuManageV1alpha1ClusterSpec
+	}{
+		{
+			description:      "check for nil data in cluster spec",
+			inputClusterSpec: nil,
+			inputRespSpec:    nil,
+			expectedSpec:     nil,
+		},
+		{
+			description: "scenario for attach cluster",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+			},
+		},
+		{
+			description: "scenario for TKG AWS spec",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgAws: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodepool{
+										AvailabilityZone: "us-west-2a",
+										InstanceType:     "m5.large",
+										NodePlacement: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodePlacement{
+											{
+												AvailabilityZone: "us-west-2a",
+											},
+										},
+										SubnetID: "default",
+										Version:  "v1.21.2+vmware.1",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgAws: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodepool{
+										AvailabilityZone: "us-east-2a",
+										InstanceType:     "m5.large",
+										NodePlacement: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodePlacement{
+											{
+												AvailabilityZone: "us-west-2a",
+											},
+										},
+										SubnetID: "default",
+										Version:  "v1.21.2+vmware.1",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgAws: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodepool{
+										AvailabilityZone: "us-west-2a",
+										InstanceType:     "m5.large",
+										NodePlacement: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodePlacement{
+											{
+												AvailabilityZone: "us-west-2a",
+											},
+										},
+										SubnetID: "default",
+										Version:  "v1.21.2+vmware.1",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			description: "scenario for TKG AWS spec topology is nil",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgAws: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodepool{
+										AvailabilityZone: "us-west-2a",
+										InstanceType:     "m5.large",
+										NodePlacement: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGAWSNodePlacement{
+											{
+												AvailabilityZone: "us-west-2a",
+											},
+										},
+										SubnetID: "default",
+										Version:  "v1.21.2+vmware.1",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: nil,
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgAws: &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsSpec{
+					Topology: nil,
+				},
+			},
+		},
+		{
+			description: "scenario for TKG vSphere spec",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGVsphereNodepool{
+										VMConfig: &nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGVsphereVMConfig{
+											CPU:       "2",
+											DiskGib:   "40",
+											MemoryMib: "8192",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGVsphereNodepool{
+										VMConfig: &nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGVsphereVMConfig{
+											CPU:       "2",
+											DiskGib:   "40",
+											MemoryMib: "8192",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGVsphereNodepool{
+										VMConfig: &nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGVsphereVMConfig{
+											CPU:       "2",
+											DiskGib:   "40",
+											MemoryMib: "8192",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			description: "scenario for TKG vSphere spec topology is nil",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									WorkerNodeCount: "1",
+									TkgVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGVsphereNodepool{
+										VMConfig: &nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGVsphereVMConfig{
+											CPU:       "2",
+											DiskGib:   "40",
+											MemoryMib: "8192",
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: nil,
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgVsphere: &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereSpec{
+					Topology: nil,
+				},
+			},
+		},
+		{
+			description: "scenario for TKGs spec",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									CloudLabels:     map[string]string{"cloud-key": "cloud-value"},
+									NodeLabels:      map[string]string{"node-key": "node-value"},
+									WorkerNodeCount: "1",
+									TkgServiceVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGServiceVsphereNodepool{
+										Class:        "test-class-spec",
+										StorageClass: "test-storage-spec",
+										Volumes: []*nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume{
+											{
+												Capacity:     4,
+												MountPath:    "/var/lib/etcd",
+												Name:         "etcd-0",
+												StorageClass: "tkgs-k8s-obj-policy",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									CloudLabels:     map[string]string{"cloud-key": "cloud-value"},
+									NodeLabels:      map[string]string{"node-key": "node-value"},
+									WorkerNodeCount: "1",
+									TkgServiceVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGServiceVsphereNodepool{
+										Class:        "test-class-spec",
+										StorageClass: "test-storage-spec",
+										Volumes: []*nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume{
+											{
+												Capacity:     4,
+												MountPath:    "/var/lib/etcd",
+												Name:         "etcd-0",
+												StorageClass: "tkgs-k8s-obj-policy",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									CloudLabels:     map[string]string{"cloud-key": "cloud-value"},
+									NodeLabels:      map[string]string{"node-key": "node-value"},
+									WorkerNodeCount: "1",
+									TkgServiceVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGServiceVsphereNodepool{
+										Class:        "test-class-spec",
+										StorageClass: "test-storage-spec",
+										Volumes: []*nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume{
+											{
+												Capacity:     4,
+												MountPath:    "/var/lib/etcd",
+												Name:         "etcd-0",
+												StorageClass: "tkgs-k8s-obj-policy",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			description: "scenario for TKGs spec topology is nil",
+			inputClusterSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{
+						NodePools: []*nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
+							{
+								Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{
+									Name:        "test",
+									Description: "testing get nodepool function",
+								},
+								Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{
+									CloudLabels:     map[string]string{"cloud-key": "cloud-value"},
+									NodeLabels:      map[string]string{"node-key": "node-value"},
+									WorkerNodeCount: "1",
+									TkgServiceVsphere: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolTKGServiceVsphereNodepool{
+										Class:        "test-class-spec",
+										StorageClass: "test-storage-spec",
+										Volumes: []*nodepoolmodel.VmwareTanzuManageV1alpha1CommonClusterTKGServiceVsphereVolume{
+											{
+												Capacity:     4,
+												MountPath:    "/var/lib/etcd",
+												Name:         "etcd-0",
+												StorageClass: "tkgs-k8s-obj-policy",
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			inputRespSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: nil,
+				},
+			},
+			expectedSpec: &clustermodel.VmwareTanzuManageV1alpha1ClusterSpec{
+				ClusterGroupName: "default",
+				TkgServiceVsphere: &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereSpec{
+					Topology: nil,
+				},
+			},
+		},
+	}
+
+	for _, each := range cases {
+		test := each
+		t.Run(test.description, func(t *testing.T) {
+			setNodepoolForClusterResource(test.inputRespSpec, test.inputClusterSpec)
+			require.Equal(t, test.expectedSpec, test.inputRespSpec)
+		})
+	}
+}

--- a/internal/resources/cluster/tkgaws/resource_tkg_aws.go
+++ b/internal/resources/cluster/tkgaws/resource_tkg_aws.go
@@ -619,8 +619,8 @@ func expandTKGAWSTopology(data []interface{}) (topology *tkgawsmodel.VmwareTanzu
 	topology = &tkgawsmodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgawsTopology{}
 
 	if v, ok := lookUpTopology[controlPlaneKey]; ok {
-		if v1, ok := v.([]interface{}); ok {
-			topology.ControlPlane = expandTKGAWSTopologyControlPlane(v1)
+		if cp, ok := v.([]interface{}); ok {
+			topology.ControlPlane = expandTKGAWSTopologyControlPlane(cp)
 		}
 	}
 
@@ -812,7 +812,11 @@ var tkgAWSNodePoolSpec = &schema.Schema{
 }
 
 func expandTKGAWSTopologyNodePool(data interface{}) (nodePools *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition) {
-	lookUpNodepool := data.(map[string]interface{})
+	lookUpNodepool, ok := data.(map[string]interface{})
+	if !ok {
+		return nodePools
+	}
+
 	nodePools = &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
 		Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{},
 		Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{},

--- a/internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go
+++ b/internal/resources/cluster/tkgservicevsphere/resource_tkg_service_vsphere.go
@@ -337,16 +337,18 @@ func expandTKGSTopology(data []interface{}) (topology *tkgservicevspheremodel.Vm
 		return topology
 	}
 
-	topology = &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{}
 	topologyData, _ := data[0].(map[string]interface{})
+	topology = &tkgservicevspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgservicevsphereTopology{}
 
 	if v, ok := topologyData[controlPlaneKey]; ok {
-		topology.ControlPlane = expandTKGSTopologyControlPlane(v.([]interface{}))
+		if cp, ok := v.([]interface{}); ok {
+			topology.ControlPlane = expandTKGSTopologyControlPlane(cp)
+		}
 	}
 
 	if v, ok := topologyData[nodePoolsKey]; ok {
-		nodePools, _ := v.([]interface{})
-		for _, np := range nodePools {
+		nodepools, _ := v.([]interface{})
+		for _, np := range nodepools {
 			topology.NodePools = append(topology.NodePools, expandTKGSTopologyNodePool(np))
 		}
 	}
@@ -366,7 +368,7 @@ func flattenTKGSTopology(topology *tkgservicevspheremodel.VmwareTanzuManageV1alp
 	nodePools := make([]interface{}, 0)
 
 	for _, nodePool := range topology.NodePools {
-		nodePools = append(nodePools, FlattenTKGSTopologyNodePool(nodePool))
+		nodePools = append(nodePools, flattenTKGSTopologyNodePool(nodePool))
 	}
 
 	flattenTopology[nodePoolsKey] = nodePools
@@ -605,7 +607,11 @@ var storageClass = &schema.Schema{
 }
 
 func expandTKGSTopologyNodePool(data interface{}) (nodePools *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition) {
-	nodePoolData := data.(map[string]interface{})
+	nodePoolData, ok := data.(map[string]interface{})
+	if !ok {
+		return nodePools
+	}
+
 	nodePools = &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
 		Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{},
 		Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{},
@@ -682,7 +688,7 @@ func expandNodePoolTKGSServiceVsphere(data []interface{}) (tkgsServiceVsphere *n
 	return tkgsServiceVsphere
 }
 
-func FlattenTKGSTopologyNodePool(nodePool *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition) (data interface{}) {
+func flattenTKGSTopologyNodePool(nodePool *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition) (data interface{}) {
 	flattenNodePool := make(map[string]interface{})
 
 	if nodePool == nil {

--- a/internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go
+++ b/internal/resources/cluster/tkgvsphere/resource_tkgvsphere.go
@@ -438,8 +438,8 @@ func expandTKGVsphereTopology(data []interface{}) (topology *tkgvspheremodel.Vmw
 	topology = &tkgvspheremodel.VmwareTanzuManageV1alpha1ClusterInfrastructureTkgvsphereTopology{}
 
 	if v, ok := lookUpTopology[controlPlaneKey]; ok {
-		if v1, ok := v.([]interface{}); ok {
-			topology.ControlPlane = expandTKGVsphereTopologyControlPlane(v1)
+		if cp, ok := v.([]interface{}); ok {
+			topology.ControlPlane = expandTKGVsphereTopologyControlPlane(cp)
 		}
 	}
 
@@ -605,7 +605,11 @@ var tkgVsphereNodePoolSpec = &schema.Schema{
 }
 
 func expandTKGVsphereTopologyNodePool(data interface{}) (nodePools *nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition) {
-	lookUpNodepool := data.(map[string]interface{})
+	lookUpNodepool, ok := data.(map[string]interface{})
+	if !ok {
+		return nodePools
+	}
+
 	nodePools = &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolDefinition{
 		Spec: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolSpec{},
 		Info: &nodepoolmodel.VmwareTanzuManageV1alpha1ClusterNodepoolInfo{},

